### PR TITLE
Run Prometheus on a different port, optionally.

### PR DESCRIPTION
**Pull request by hawkowl**
This should allow us to be resistant to Twisted melting down meaning we don't get metrics, in theory.

### CODE REVIEW

**Comment by hawkowl** [comment]
@erikjohnston this is based off the other PR, so will become much smaller when that one merges :)

**Comment by erikjohnston** [comment]
I think my preference would be to make it configurable via the `listeners` config rather than reusing the old `metrics_port` config options. Maybe something like:

```yaml
listeners:
- port: 9000
  type: metrics
```

Also, we need to update each component rather than just the main one, i.e. everything else in `synapse/app/*` (yes, the duplication there makes me sad too)

**Comment by turt2live** [comment]
If the configuration changes, please yell loudly in the changelog too 😇

**Comment by erikjohnston** [comment]
@turt2live I imagine we'd still support hosting the metrics endpoint in the twisted reactor as well, so everything should be backwards compat :)

**Comment by hawkowl** [comment]
@erikjohnston New docs, added the config to all the apps, listeners config added.

File: synapse/app/_base.py [hunk 1]
```diff
@@ -124,6 +124,19 @@ def quit_with_error(error_string):
...
+def listen_metrics(bind_addresses, port):
+    """
+    Start Prometheus metrics server.
+    """
+    from synapse.metrics import RegistryProxy
+    from prometheus_client import start_http_server
+
+    for host in bind_addresses:
+        reactor.callInThread(start_http_server, int(port),
+                             addr=host, registry=RegistryProxy)
```
Comment (erikjohnston) [inline synapse/app/_base.py]: `start_http_server` should spawn its own thread, so is it necessary to `callInThread`?

File: synapse/config/server.py [hunk 2]
```diff
@@ -136,21 +136,6 @@ def read_config(self, config):
...
-                "bind_addresses": [config.get("metrics_bind_host", "127.0.0.1")],
-                "tls": False,
-                "type": "http",
-                "resources": [
-                    {
-                        "names": ["metrics"],
-                        "compress": False,
-                    },
-                ]
-            })
```
Comment (erikjohnston) [inline synapse/config/server.py]: Do we not need to keep this for backwards compat for people using `metrics_port`?

File: synapse/app/_base.py (hunk 1 above)
Comment (hawkowl) [reply synapse/app/_base.py]: I had issues with it jamming Twisted, for some reason, and this made it work fine. ~the mysteries of threading~

**Comment by hawkowl** [comment]
@erikjohnston I put back the old metrics_port, and raised a warning saying it's deprecated, with a link to the new docs. I couldn't find any better globally accessible link than just the file in the repo, though...

File: docs/metrics-howto.rst [hunk 3]
```diff
@@ -31,7 +53,40 @@ How to monitor Synapse metrics using Prometheus
...
+
+
+Removal of deprecated metrics & time based counters becoming histograms in 0.31.0
+---------------------------------------------------------------------------------
+
+The duplicated metrics deprecated in Synapse 0.27.0 have been removed.
+
+All time duration-based metrics have been changed to be seconds. This affects:
+
+================================
```
Comment (richvdh) [inline docs/metrics-howto.rst]: this table doesn't appear to be working. Please take a look at the rendered version: https://github.com/matrix-org/synapse/blob/develop/docs/metrics-howto.rst#removal-of-deprecated-metrics--time-based-counters-becoming-histograms-in-0310

File: docs/metrics-howto.rst [hunk 4]
```diff
@@ -1,25 +1,47 @@
...
+   be available at ``/``.
 
-3. Add a prometheus target for synapse.
+   Add a new listener to homeserver.yaml::
+
+     listeners:
+       - type: metrics
+         port: 9000
+         bind_addresses:
+           - '0.0.0.0'
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply docs/metrics-howto.rst]: SocketServer (from the stdlib that Prometheus Client uses) only supports IPv4 out of the box, so it refuses IPv6 addresses. :(

[the remark this answers is not in the document]
Comment (richvdh) [reply docs/metrics-howto.rst]: whaaa

**Pull request merged**